### PR TITLE
refactor: extract shared types and utility functions to mdl/types

### PR DESCRIPTION
**Pull request by retran**
## Why

The executor layer (`mdl/`) imports `sdk/mpr` for shared type definitions — `EdmxDocument`, `JavaAction`, `NavigationProfile`, `JsonStructure`, and dozens more. This coupling means any package that needs these types must pull in the entire MPR reader/writer stack, preventing alternative backend implementations and making the `mdl/` subtree unsuitable for lightweight compilation targets.

This PR extracts all shared types into a new `mdl/types` package that has no dependency on `sdk/mpr`.

## What changed

**New packages:**
- `mdl/types/` — 9 files containing all shared domain types previously defined in `sdk/mpr/reader_types.go`, `sdk/mpr/edmx.go`, `sdk/mpr/asyncapi.go`, `sdk/mpr/utils.go`, and `sdk/mpr/writer_jsonstructure.go`
- `mdl/backend/mpr/convert.go` — bidirectional conversion layer between `sdk/mpr` reader/writer types and `mdl/types`

**Extracted types and functions:**
- **id.go** — `GenerateID`, `GenerateDeterministicID`, `BlobToUUID`, `UUIDToBlob`, `ValidateID`, `Hash`
- **infrastructure.go** — `MPRVersion`, `ProjectVersion`, `FolderInfo`, `UnitInfo`, `RenameHit`, `RawUnit`, `RawUnitInfo`, `RawCustomWidgetType`, `EntityMemberAccess`, `EntityAccessRevocation`
- **java.go** — `JavaAction`, `JavaScriptAction`
- **mapping.go** — `JsonStructure`, `JsonElement`, `ImageCollection`, `Image`
- **navigation.go** — `NavigationDocument`, `NavigationProfile`, `NavHomePage`, `NavRoleBasedHome`, `NavMenuItem`, `NavOfflineEntity`, plus spec types
- **edmx.go** — Full EDMX/CSDL parser with all `Edm*` types and `ParseEdmx`
- **asyncapi.go** — Full AsyncAPI parser with all `AsyncAPI*` types and `ParseAsyncAPI`
- **json_utils.go** — `PrettyPrintJSON`, `BuildJsonElementsFromSnippet`, datetime normalization

**Migration (65 files):** All executor handlers, catalog builders, backend interfaces, and mock implementations updated from `mpr.*` to `types.*` — purely mechanical import swaps with no logic changes.

**Backward compatibility:** `sdk/mpr` retains thin delegation wrappers and returns `types.*` directly, so external callers are unaffected.

## Stack

PR 1/5 in the shared-types refactoring stack.
Merge order: #232 → **this** → #236 → #237 → #238 → #239

### CODE REVIEW

File: sdk/mpr/utils.go [hunk 1]
```diff
@@ -3,66 +3,48 @@
...
+	return primitive.Binary{
+		Subtype: 0x00,
+		Data:    blob,
+	}
 }
 
 // BsonBinaryToID converts a BSON binary value to a UUID string.
 func BsonBinaryToID(bin primitive.Binary) string {
-	return BlobToUUID(bin.Data)
+	return types.BlobToUUID(bin.Data)
```
[the remark this answers is not in the document]
Comment (retran) [reply sdk/mpr/utils.go:32]: Good catch — the `mdl/bsonutil` package is introduced in PR #236 (next in the stack). Updated the PR description to remove the premature mention. Thanks!

File: mdl/executor/cmd_pages_builder.go [hunk 2]
```diff
@@ -305,7 +306,7 @@ func (pb *pageBuilder) resolveFolder(folderPath string) (model.ID, error) {
 			currentContainerID = newFolderID
 
 			// Add to cache
-			pb.foldersCache = append(pb.foldersCache, &mpr.FolderInfo{
+			pb.foldersCache = append(pb.foldersCache, &types.FolderInfo{
 				ID:          newFolderID,
 				ContainerID: currentContainerID,
 				Name:        part,
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/executor/cmd_pages_builder.go:313]: Real bug, good find! This was pre-existing (same pattern before this PR's type swap), but fixed now — captured `parentContainerID` before updating `currentContainerID`. Fixed in dbfb477.

File: mdl/types/id.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+	"fmt"
+	"strings"
+)
+
+// GenerateID generates a new unique UUID v4 for model elements.
+func GenerateID() string {
+	b := make([]byte, 16)
+	_, _ = rand.Read(b)
+	b[6] = (b[6] & 0x0f) | 0x40 // Version 4
+	b[8] = (b[8] & 0x3f) | 0x80 // Variant is 10
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/types/id.go:20]: Agreed — `crypto/rand.Read` failure should not be silent. Changed to panic with a clear message since this is a system-level failure (no entropy source). Fixed in dbfb477.

File: sdk/mpr/utils.go [hunk 4]
```diff
@@ -3,66 +3,48 @@
...
 func IDToBsonBinary(id string) primitive.Binary {
-	return idToBsonBinary(id)
+	blob := types.UUIDToBlob(id)
+	if blob == nil || len(blob) != 16 {
+		blob = types.UUIDToBlob(types.GenerateID())
+	}
+	return primitive.Binary{
+		Subtype: 0x00,
+		Data:    blob,
+	}
```
[the remark this answers is not in the document]
Comment (retran) [reply sdk/mpr/utils.go]: Fixed — `IDToBsonBinary` now delegates to `idToBsonBinary`, single implementation. Fixed in dbfb477.

File: sdk/mpr/writer_imagecollection.go [hunk 5]
```diff
@@ -3,13 +3,14 @@
...
 	"go.mongodb.org/mongo-driver/bson"
 	"go.mongodb.org/mongo-driver/bson/primitive"
 )
 
 // CreateImageCollection creates a new empty image collection unit in the MPR.
-func (w *Writer) CreateImageCollection(ic *ImageCollection) error {
+func (w *Writer) CreateImageCollection(ic *types.ImageCollection) error {
 	if ic.ID == "" {
 		ic.ID = model.ID(generateUUID())
 	}
```
[the remark this answers is not in the document]
Comment (retran) [reply sdk/mpr/writer_imagecollection.go:15]: Good catch — added type aliases for `ImageCollection`, `Image`, and all other extracted types in `sdk/mpr/reader_types.go`. External callers can still reference `mpr.ImageCollection` etc. Fixed in 62ce850.

File: sdk/mpr/parser_misc.go [hunk 6]
```diff
@@ -91,7 +92,7 @@ func (r *Reader) parseJavaAction(unitID, containerID string, contents []byte) (*
 		return nil, fmt.Errorf("failed to unmarshal BSON: %w", err)
 	}
 
-	ja := &JavaAction{}
+	ja := &types.JavaAction{}
 	ja.ID = model.ID(unitID)
```
[the remark this answers is not in the document]
Comment (retran) [reply sdk/mpr/parser_misc.go:95]: Added `type JavaAction = types.JavaAction` and `type JavaScriptAction = types.JavaScriptAction` aliases (plus all other extracted types) in `sdk/mpr/reader_types.go`. Fixed in 62ce850.

File: mdl/types/id.go [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+// UUIDToBlob converts a UUID string to a 16-byte blob in Microsoft GUID format.
+func UUIDToBlob(uuid string) []byte {
+	if uuid == "" {
+		return nil
+	}
+	var clean strings.Builder
+	for _, c := range uuid {
+		if c != '-' {
+			clean.WriteString(string(c))
+		}
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/types/id.go:61]: Fixed — `UUIDToBlob` now uses `clean.Grow(32)` + `clean.WriteByte()` instead of per-rune string allocation. Fixed in 349b579.

File: mdl/types/asyncapi.go [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+	doc := &AsyncAPIDocument{
+		Version:     raw.AsyncAPI,
+		Title:       raw.Info.Title,
+		DocVersion:  raw.Info.Version,
+		Description: raw.Info.Description,
+	}
+
+	// Resolve messages from components
+	for name, msg := range raw.Components.Messages {
+		resolved := &AsyncAPIMessage{
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/types/asyncapi.go]: Fixed — sorted map iteration for Messages, Channels, and schema Properties in `ParseAsyncAPI`. Fixed in 349b579.

File: sdk/mpr/writer_imagecollection.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (retran) [reply sdk/mpr/writer_imagecollection.go:15]: Good point — switched all `sdk/mpr` exported signatures and internal functions to use local type aliases (`ImageCollection`, `JsonStructure`, `JsonElement`, `JavaAction`, etc.) instead of `types.*`. The `mdl/types` import is now only used in delegation wrappers. Fixed in 8282e92.

**Comment by retran** [comment]
Thanks for the reviews. RE the conversion function parameter types flagged in the third review: these are not bugs. The `sdk/mpr` reader now returns `*types.FolderInfo` directly (via type alias `FolderInfo = types.FolderInfo`), so the convert functions correctly accept `[]*types.FolderInfo`. The code compiles and tests pass. The passthrough converters exist for future-proofing in case the types diverge.

File: mdl/types/id.go [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+	if uuid == "" {
+		return nil
+	}
+	var clean strings.Builder
+	clean.Grow(32)
+	for _, c := range uuid {
+		if c != '-' {
+			clean.WriteByte(byte(c))
+		}
+	}
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/types/id.go:62]: Good catch — switched from rune iteration to byte iteration (`for i := 0; i < len(uuid); i++`), so multi-byte UTF-8 stays intact and hex decoding correctly rejects non-ASCII input. Fixed in 2df1c83.

File: sdk/mpr/writer_imagecollection_test.go [hunk 10]
```diff
@@ -5,12 +5,13 @@ package mpr
...
+	"github.com/mendixlabs/mxcli/mdl/types"
 	"github.com/mendixlabs/mxcli/model"
 	"go.mongodb.org/mongo-driver/bson"
 )
 
 func TestSerializeImageCollection_EmptyImages(t *testing.T) {
-	ic := &ImageCollection{
+	ic := &types.ImageCollection{
 		BaseElement: model.BaseElement{ID: "ic-test-1"},
 		ContainerID: model.ID("module-id-1"),
```
[the remark this answers is not in the document]
Comment (retran) [reply sdk/mpr/writer_imagecollection_test.go:15]: Fixed — test now uses `&ImageCollection{...}` via the local type alias, removed `mdl/types` import. Fixed in 2df1c83.

**Comment by retran** [comment]
RE the "Critical Issues" about conversion function signatures (comment from 18:10:13) and the missing tests note (comment from 18:15:46):

**Conversion signatures are correct — not bugs.** The `sdk/mpr` package defines type aliases:

```go
// sdk/mpr/reader_types.go
type FolderInfo = types.FolderInfo
type JavaAction = types.JavaAction
type JsonStructure = types.JsonStructure
// ... (16 aliases total)
```

Because these are `=` aliases (not new types), `[]*types.FolderInfo` and `[]*mpr.FolderInfo` are **the same Go type**. The SDK reader methods already return `*types.FolderInfo` through the alias. The convert functions accept `types.*` directly because that IS `mpr.*`. Same applies to the "unconvert" functions — `*types.JsonStructure` IS `*mpr.JsonStructure`.

This is now proven by `mdl/backend/mpr/convert_test.go` (added in f2c8e12), which contains compile-time type identity assertions:

```go
var _ *types.JavaAction = new(mpr.JavaAction)
var typesSlice []*types.FolderInfo
var mprSlice []*mpr.FolderInfo = typesSlice  // compiles because same type
```

**Tests added** in f2c8e12 (+1094 lines across 5 test files):
- `mdl/types/id_test.go` — UUID generation, BlobToUUID/UUIDToBlob roundtrip, ValidateID, Hash
- `mdl/types/json_utils_test.go` — PrettyPrintJSON, datetime normalization, BuildJsonElementsFromSnippet edge cases
- `mdl/types/edmx_test.go` — OData4 parsing, enums, capability annotations, FindEntityType, ResolveNavType
- `mdl/types/asyncapi_test.go` — AsyncAPI parsing, sorted channels/messages, FindMessage
- `mdl/backend/mpr/convert_test.go` — type alias identity proofs, field preservation across alias boundary

Also fixed a real bug caught by the new tests: `normalizeDateTimeValue` was matching `-` in the date portion instead of the timezone suffix.

File: mdl/types/json_utils.go [hunk 11]
```diff
@@ -0,0 +1,378 @@
...
+		// No fractional part — insert .0000000 before timezone suffix.
+		// Search from index 19+ to avoid matching the '-' in the date portion (YYYY-MM-DD).
+		if len(s) > 19 {
+			if idx := strings.IndexAny(s[19:], "Z+-"); idx >= 0 {
+				pos := 19 + idx
+				return s[:pos] + ".0000000" + s[pos:]
+			}
+		}
+		return s
+	}
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/types/json_utils.go:47]: Fixed — `normalizeDateTimeValue` now handles the `len(s) == 19` case (no fractional, no timezone) by appending `.0000000` at the end. Added a dedicated test `TestNormalizeDateTimeValue_NoFractionalNoTimezone`. Fixed in a807b77.

File: mdl/types/json_utils.go [hunk 12]
```diff
@@ -0,0 +1,378 @@
...
+			primitiveType = "DateTime"
+			v = normalizeDateTimeValue(v)
+		}
+		return buildValueElement(exposedName, path, primitiveType, fmt.Sprintf("%q", v))
+	case float64:
+		// Check the raw JSON text for a decimal point — Go's %v drops ".0" from 41850.0
+		if v == math.Trunc(v) && !strings.Contains(trimmed, ".") {
+			return buildValueElement(exposedName, path, "Integer", fmt.Sprintf("%v", int64(v)))
+		}
+		return buildValueElement(exposedName, path, "Decimal", fmt.Sprintf("%v", v))
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/types/json_utils.go:237]: Fixed — added `v >= -(1<<53) && v <= (1<<53)` safe integer range guard. Values outside this range fall through to Decimal, avoiding silent precision loss. Fixed in a807b77.

File: mdl/types/json_utils.go [hunk 13]
```diff
@@ -0,0 +1,378 @@
...
+func (b *snippetBuilder) resolveExposedName(jsonKey string) string {
+	if b.customNameMap != nil {
+		if custom, ok := b.customNameMap[jsonKey]; ok {
+			return custom
+		}
+	}
+	name := capitalizeFirst(jsonKey)
+	if reservedExposedNames[name] {
+		return "_" + jsonKey
+	}
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/types/json_utils.go:127]: Fixed — removed "Name" from the comment. The map only contains `Id` and `Type`; the comment now matches. Fixed in a807b77.

File: mdl/types/json_utils.go [hunk 14]
```diff
@@ -0,0 +1,378 @@
...
+			arrayElem.Children = append(arrayElem.Children, wrapper)
+		}
+	}
+
+	return arrayElem
+}
+
+// singularize returns a simple singular form by stripping trailing "s".
+// Handles common cases: Tags→Tag, Items→Item, Addresses→Addresse.
+func singularize(s string) string {
```
[the remark this answers is not in the document]
Comment (retran) [reply mdl/types/json_utils.go]: Clarified — the comment now explicitly states this is intentionally naive and matches Studio Pro's behavior: "Known-incorrect for some words (e.g. Addresses→Addresse) — this matches Studio Pro's behavior." Fixed in a807b77.

**Comment by ako** [comment]
## Code Review

### Overview

This is a significant architectural refactoring (113 files, +7647/-2090 lines) that decouples the `mdl/` layer from `sdk/mpr` by extracting shared domain types into a new `mdl/types` package. The goal — allowing lightweight compilation targets and alternative backend implementations without pulling in the MPR reader/writer stack — is sound and well-motivated. The approach is solid overall, with a few maintenance risks worth flagging.

---

### Positive Aspects

- **Type alias strategy is the right call.** Using `type JavaAction = types.JavaAction` in `sdk/mpr/reader_types.go` preserves 100% backward compatibility: existing callers don't break, slice assignments are interchangeable, and there's a single source of truth.
- **Extraction is complete.** No `mdl/` packages were left with orphaned `sdk/mpr` imports for the extracted types. The boundary is clean.
- **Conversion layer is thorough.** All read-path (sdk→mdl) and write-path (mdl→sdk) conversions are accounted for, including recursive structures (`NavMenuItem`, `JsonElement`).
- **`convert_test.go` proves the alias strategy works** at compile time via assignment interchangeability tests. That's exactly the kind of test this pattern needs.

---

### Issues & Concerns

**1. Spec types are duplicated, not aliased — medium maintenance risk**

`NavigationProfileSpec`, `NavHomePageSpec`, and `NavMenuItemSpec` are defined independently in both `mdl/types` and `sdk/mpr`. Unlike all other extracted types, these are **not** type aliases — the `unconvert` functions copy between two distinct struct definitions.

This means:
- A new field added to `mdl/types.NavigationProfileSpec` won't automatically appear in `mpr.NavigationProfileSpec`
- BSON/JSON tag changes must be made in two places
- The compiler won't catch the divergence

**Suggestion:** Either alias these too (`type NavigationProfileSpec = types.NavigationProfileSpec` in `sdk/mpr`) or add a compile-time assertion that the two types are structurally identical.

**2. Conversion functions are untested — medium risk**

`convert_test.go` verifies that type aliases compile correctly and that fields survive assignment. But none of the actual **conversion functions** (`convertNavDoc`, `convertJsonStructure`, `convertNavMenuItem`, `convertImageCollection`, etc.) are exercised with field-preservation round-trip tests.

If a field is added to `mdl/types.NavigationProfile` but forgotten in `convertNavProfile()`, the test suite won't catch it.

**Suggestion:** Add at least one round-trip test per conversion family — construct a fully-populated `sdk/mpr` struct, convert to `mdl/types`, verify all fields survive. The existing `TestNavigationDocumentFieldPreservation` does this for the *alias* case but not for the *conversion function* case.

**3. No tests for reverse (unconvert) paths**

The `unconvertNavProfileSpec`, `unconvertEntityMemberAccessSlice`, `unconvertJsonStructure`, etc. have zero test coverage. These are write-path critical paths — a bug here silently produces bad BSON writes.

**Suggestion:** Add unit tests for each `unconvert*` function, at minimum verifying fields aren't dropped.

**4. Manual field copying is a long-term maintenance burden**

The conversion functions do explicit field-by-field copies. Any struct evolution (new field added, field renamed) requires updating the copy in `convert.go`. There's no mechanism to catch drift.

This is somewhat unavoidable given the approach, but it's worth flagging. A `go vet`-style check or a compile-time struct comparison helper (e.g., asserting field counts match) would help.

---

### Checklist Against Project Standards

| Item | Status |
|---|---|
| Import cleanup complete | ✅ No orphaned `sdk/mpr` imports in `mdl/` |
| Backward compatibility | ✅ Type aliases preserve all existing callers |
| Each commit does one thing | ✅ Tightly scoped to the extraction |
| Test coverage for new packages | ⚠️ Type alias tests present; conversion functions and reverse paths untested |
| Spec type duplication | ⚠️ `NavigationProfileSpec` etc. duplicated, not aliased |

---

### Verdict

**Approve with suggestions.** The architectural goal is correct, the extraction is complete, and the alias strategy is the right implementation choice. The two concrete risks are (a) the duplicated Spec types diverging silently and (b) conversion functions having no field-preservation tests. Neither blocks merging, but both should be tracked as follow-up work — the longer they go without tests, the more expensive they become to retrofit.
